src/transformers/models/whisper/generation_whisper.py

```diff
@@ -1016,94 +1072,211 @@ def _set_language_and_task(language, task, is_multilingual, generation_config):
+                    "(You should just add it to the generation config)"
+                )
+
+            lang_id = generation_config.lang_to_id[language_token]
+
+            # if language is defined it'll overwrite language ids that might have already been defined via the generation_config
+            replace_or_add(init_tokens, lang_id, generation_config.lang_to_id.values())
+        elif hasattr(generation_config, "lang_to_id") and is_lang_id_undefined:
+            # language is not defined or intentially set to `None` to trigger language detection
+            lang_ids = self.detect_language(
+                input_features=input_features,
+                encoder_outputs=kwargs.get("encoder_outputs", None),
+                generation_config=generation_config,
+                num_segment_frames=num_segment_frames,
+            )
+
+            if torch.unique(lang_ids).shape[0] > 1:
+                raise ValueError(
+                    "Multiple languages detected when trying to guess the target language for transcription. It is currently not supported to transcribe to different languages in a single batch. Please make sure to either force a single language by passing `languag='...'` or make sure all input audio is of the same language."
+                )
+
+            lang_id = lang_ids[0].item()
+
+            # append or replace lang_id to init_tokens
+            if len(init_tokens) > 1:
+                init_tokens[1] = lang_id
+            else:
+                init_tokens.append(lang_id)
+
+        if task is not None:
+            if task in TASK_IDS:
+                init_tokens.append(generation_config.task_to_id[generation_config.task])
+                task_id = generation_config.task_to_id[generation_config.task]
+
+                # if task is defined it'll overwrite task ids that might have already been defined via the generation_config
+                replace_or_add(init_tokens, task_id, generation_config.task_to_id.values())
+            else:
+                raise ValueError(f"The `{task}`task is not supported. The task should be one of `{TASK_IDS}`")
+        elif language is not None and hasattr(generation_config, "task_to_id"):
+            # if language is defined, but no task id is in `init_tokens`, default to transcribe
+            if not any(i in init_tokens for i in generation_config.task_to_id.values()):
+                init_tokens.append(generation_config.task_to_id["transcribe"])
+
+        if (
+            not generation_config.return_timestamps
+            and hasattr(generation_config, "no_timestamps_token_id")
+            and init_tokens[-1] != generation_config.no_timestamps_token_id
+        ):
+            init_tokens.append(generation_config.no_timestamps_token_id)
+        elif generation_config.return_timestamps and init_tokens[-1] == generation_config.no_timestamps_token_id:
```

Comment:
(e.g. here we always override the forced ids with the setting of `return_timestamps`)